crates/storage/src/blueprint/sparse.rs

```diff
@@ -484,237 +484,483 @@ where
     }
 }
 
-/// The macro that generates SMT storage tests for the table with [`crate::structured_storage::test::InMemoryStorage`].
 #[cfg(feature = "test-helpers")]
-#[macro_export]
-macro_rules! root_storage_tests {
-    ($table:ident, $metadata_table:ident, $current_key:expr, $foreign_key:expr, $generate_key:ident, $generate_value:ident) => {
-        paste::item! {
-        #[cfg(test)]
-        mod [< $table:snake _root_tests >] {
-            use super::*;
-            use $crate::{
-                structured_storage::test::InMemoryStorage,
-                transactional::WriteTransaction,
-                StorageAsMut,
-            };
-            use $crate::rand::{
-                rngs::StdRng,
-                SeedableRng,
+#[allow(warnings)]
+/// Test module for root storage tests.
+pub mod root_storage_tests {
+    use fuel_vm_private::{
+        fuel_merkle::sparse::{
+            self,
+            proof::Proof,
+            MerkleTree,
+            MerkleTreeKey,
+        },
+        fuel_storage::{
+            Mappable,
+            MerkleRoot,
+            StorageAsMut,
+        },
+    };
+    use rand::{
+        rngs::StdRng,
+        SeedableRng,
+    };
+
+    use crate::{
+        blueprint::BlueprintMutate,
+        codec::{
+            Decode,
+            Encode,
+            Encoder,
+        },
+        structured_storage::{
+            test::InMemoryStorage,
+            StructuredStorage,
+            TableWithBlueprint,
+        },
+        tables::merkle::SparseMerkleMetadata,
+        transactional::{
+            StorageTransaction,
+            WriteTransaction,
+        },
+    };
+
+    use super::{
+        PrimaryKey,
+        Sparse,
+    };
+
+    /// Provides root storage tests for SMT storage tables.
+    pub trait RootStorageTests: SmtTableWithBlueprint
+    where
+        Self::OwnedValue: PartialEq + core::fmt::Debug,
+        Self::Column: PartialEq,
+        <Self::Metadata as Mappable>::Key: Sized,
+
+        for<'a, 'b> <Self::Metadata as TableWithBlueprint>::Blueprint: BlueprintMutate<
+            Self::Metadata,
+            StructuredStorage<
+                &'a mut StorageTransaction<&'b mut InMemoryStorage<Self::Column>>,
+            >,
+        >,
+        for<'a> <Self::Metadata as TableWithBlueprint>::Blueprint: BlueprintMutate<
+            Self::Metadata,
+            StorageTransaction<&'a mut InMemoryStorage<Self::Column>>,
+        >,
+
+        for<'a, 'b> <Self::Nodes as TableWithBlueprint>::Blueprint: BlueprintMutate<
```

Comment:
helper type would be nice here, as mentioned prev

Reply:
https://github.com/FuelLabs/fuel-core/pull/2914/commits/e1ffdb9556b262eceddad322ebe91cd546001247